File: proyectoFirebase/src/app/servicios/navigation.service.ts
/*
BREADCRUMBS NAVIGATION
Service to manage navigation state and breadcrumbs across the app.
Allows setting selected patient context and building dynamic breadcrumbs.
Enables navigation with context preservation.

*/

import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject } from 'rxjs';

export interface Breadcrumb {
  label: string;
  path: string;
  params?: any;
}

export interface NavigationState {
  selectedPatient?: any;
  fromPage?: string;
  breadcrumbs: Breadcrumb[];
}

@Injectable({
  providedIn: 'root'
})
export class NavigationService {
  private navigationState = new BehaviorSubject<NavigationState>({
    breadcrumbs: []
  });

  public navigationState$ = this.navigationState.asObservable();

  constructor(private router: Router) {}

  // Set the selected patient and update breadcrumbs
  setSelectedPatient(patient: any, fromPage?: string) {
    const currentState = this.navigationState.value;
    const newState: NavigationState = {
      ...currentState,
      selectedPatient: patient,
      fromPage: fromPage,
      breadcrumbs: this.buildBreadcrumbs(patient, fromPage)
    };
    
    this.navigationState.next(newState);
  }

  // Get current navigation state
  getCurrentState(): NavigationState {
    return this.navigationState.value;
  }

  // Get selected patient
  getSelectedPatient() {
    return this.navigationState.value.selectedPatient;
  }

  // Build breadcrumbs based on current context
  private buildBreadcrumbs(patient?: any, fromPage?: string): Breadcrumb[] {
    const breadcrumbs: Breadcrumb[] = [
      { label: 'Inicio', path: '/inicio' },
      { label: 'Pacientes', path: '/listadoPacientes' }
    ];

    if (patient) {
      breadcrumbs.push({
        label: patient.nombre || patient.patient_name || 'Paciente',
        path: '/perfilPaciente',
        params: { patient }
      });
    }

    return breadcrumbs;
  }

  // Navigate with breadcrumb context
  navigateWithBreadcrumb(path: string, patient?: any) {
    if (patient) {
      this.setSelectedPatient(patient, path);
    }
    
    this.router.navigate([path], {
      state: { 
        patient: patient,
        navigationState: this.getCurrentState()
      }
    });
  }

  // Execute breadcrumb navigation
  executeBreadcrumbNavigation(breadcrumb: Breadcrumb) {
    if (breadcrumb.params?.patient) {
      this.navigateWithBreadcrumb(breadcrumb.path, breadcrumb.params.patient);
    } else {
      this.router.navigate([breadcrumb.path]);
    }
  }

  // Clear navigation state
  clearNavigationState() {
    this.navigationState.next({ breadcrumbs: [] });
  }

  // Get breadcrumbs for medical record page
  getMedicalRecordBreadcrumbs(patient?: any): Breadcrumb[] {
    const breadcrumbs = this.buildBreadcrumbs(patient);
    breadcrumbs.push({
      label: 'Ficha Médica',
      path: '/fichaMedica',
      params: { patient }
    });
    return breadcrumbs;
  }
}

File: flutter_app2/lib/services/navigation_service.dart
import '../patients.dart';

class Breadcrumb {
  final String label;
  final String route;
  final Object? arguments;
  final bool isActive;

  const Breadcrumb({
    required this.label,
    required this.route,
    this.arguments,
    this.isActive = false,
  });

  Breadcrumb copyWith({
    String? label,
    String? route,
    Object? arguments,
    bool? isActive,
  }) {
    return Breadcrumb(
      label: label ?? this.label,
      route: route ?? this.route,
      arguments: arguments ?? this.arguments,
      isActive: isActive ?? this.isActive,
    );
  }
}

class NavigationService {
  NavigationService._();

  static final NavigationService instance = NavigationService._();

  Paciente? _selectedPatient;

  Paciente? get selectedPatient => _selectedPatient;

  void setSelectedPatient(Paciente? patient) {
    _selectedPatient = patient;
  }

  List<Breadcrumb> buildInicioBreadcrumbs({String inicioLabel = 'Inicio'}) {
    return [
      Breadcrumb(label: inicioLabel, route: '/inicio', isActive: true),
    ];
  }

  List<Breadcrumb> buildPatientListBreadcrumbs({
    String inicioLabel = 'Inicio',
    String patientsLabel = 'Pacientes',
  }) {
    return [
      Breadcrumb(label: inicioLabel, route: '/inicio'),
      Breadcrumb(
        label: patientsLabel,
        route: '/pacientes',
        isActive: true,
      ),
    ];
  }

  List<Breadcrumb> buildPatientProfileBreadcrumbs(
    Paciente? patient, {
    String inicioLabel = 'Inicio',
    String patientsLabel = 'Pacientes',
  }) {
    final displayName = patient == null || patient.nombrePaciente.isEmpty
        ? 'Paciente'
        : patient.nombrePaciente;
    return [
      Breadcrumb(label: inicioLabel, route: '/inicio'),
      Breadcrumb(label: patientsLabel, route: '/pacientes'),
      Breadcrumb(
        label: displayName,
        route: '/perfilPaciente',
        arguments: patient,
        isActive: true,
      ),
    ];
  }

  List<Breadcrumb> buildMedicalRecordBreadcrumbs(
    Paciente? patient, {
    String inicioLabel = 'Inicio',
    String patientsLabel = 'Pacientes',
    String recordLabel = 'Ficha Médica',
  }) {
    final crumbs = buildPatientProfileBreadcrumbs(
      patient,
      inicioLabel: inicioLabel,
      patientsLabel: patientsLabel,
    );
    crumbs.add(
      Breadcrumb(
        label: recordLabel,
        route: '/fichaMedica',
        arguments: patient,
        isActive: true,
      ),
    );
    return crumbs;
  }
}
